_posts/2019/2019-08-20-end-to-end-tests-with-geb-and-spock.md

```diff
@@ -0,0 +1,154 @@
+
+I will write about the motivation to start a test project, the goals that we set out to achieve, and some common problems
+and their solutions.
+
+
+## Motivation & Goals
+
+Since several years Selenium has been used at ePages. A framework was developed, with all the good intentions of categorising,
+structuring and maintaining the tests. But in the context of our third-party integrations this has been notoriously difficult.
+
+So after some discussions, deliberations, and a few Git branches later, we decided to start a new test project for this purpose,
+to be able to work on this independently, without any dependencies to other teams or existing code.
+
+We specifically did not want to create a new framework, to replace the old one.
+We wanted a way to simply and effectively develop end-to-end tests, and to be able to stabilise them.
+Last but not least we wanted to make use of Selenium Grid to parallelise execution of the tests, in order to improve the feedback
+cycle.
+
+[Geb](https://gebish.org/), [Spock](http://spockframework.org/) and [Gradle](https://gradle.org/) are a very good fit to approach
+this problem. There are some considerations to cut back on using the Groovy language, but this was not really a concern for us,
+because Groovy is a very easy to learn languages, and it is mostly backwards compatible to Java.
+
+<img alt="Image of Spock the Star Trek character" src="/assets/img/pages/blog/images/spock-geb-zalenium.png" class="center-image"/>
+<style type="text/css">
+.center-image {
+    height: 320px;
+    width: auto;
+    display: block;
+    margin-left: auto;
+    margin-right: auto;
+}
+</style>
+
+## Stability testing
+
+End-to-end tests generally have a bad reputation, because it is difficult to make them reliable enough to be useful. This was not
+different in our case.
+But we realised that if we wanted to improve the situation with our own tests, we would need to run them many times, regularly.
+And we would need to debug and understand every problem that has more than negligible chance of occurring.
+For this purpose we used a task rule in Gradle to create the desired amount of test tasks, i.e. repeated executions, with an
+optional test class filter in case we want to restrict execution to certain tests. After making a change to remedy a synchronisation
+problem, we would typically run the affected test between 20 and 100 times.
+
+The harder part is of course debugging the individual errors, especially if they are difficult to reproduce. Most of the
+failures in this area are synchronisation problems between the browser and the test, caused by the high complexity of our matured frontend.
+In many cases the only way to synchronise the UI with the test, is to wait for some unambiguous condition, that is the result
+of some interaction. This can be tedious at times, but it is also the general case, for example when dealing with a single page
```

Comment:
This can be tedious at times, but ~~it~~ is also the **usual** ~~general~~ (?!) case. For example, when dealing with a single page application, ~~where~~ Selenium is unable to....

Reply:
Es ist in der Tat der "allgemeine" Fall, würde ich sagen, aus technischer Sicht, aber nicht etwa der gewöhnlichste und schon gar nicht der häufigste ;)

Bin aber grade dran an dem Satz.